chore: Perform feature/config condition checks at request time

benjreinhart opened the pull request:
### SUMMARY

This builds on https://github.com/apache/superset/pull/14568 and https://github.com/apache/superset/pull/14650 by swapping out the remaining occurrences of code that conditionally registers resources during startup. It should now be possible to flip config/feature values on/off without restarting the server.

### TEST PLAN

I enabled some disabled tests and added many more. I also tested many of these endpoints manually.

cc @robdiciuccio @craig-rueda @dpgaspar

benjreinhart committed:
chore: conditional Home link rendering

benjreinhart committed:
chore: conditional RowLevelSecurity rendering

benjreinhart committed:
chore: Conditional KV rendering

benjreinhart committed:
chore: Conditional TagView rendering

benjreinhart committed:
chore: Conditional import dashboards link

benjreinhart committed:
chore: Conditional upload csv/excel links

benjreinhart committed:
chore: Conditional log api and view rendering

benjreinhart committed:
chore: Conditionally render email schedules

benjreinhart committed:
chore: Conditionally render alert views

benjreinhart committed:
chore: Conditionally render alerts/reports

benjreinhart committed:
chore: Conditionally render access requests

benjreinhart committed:
chore: Conditionally render druid views

benjreinhart committed:
Remove unnecessary folder

benjreinhart committed:
Consistent naming

benjreinhart committed:
Cleanup

benjreinhart committed:
Remove object from class

benjreinhart committed:
Clean up test file

benjreinhart committed:
Clean up test file

benjreinhart committed:
Fix lint error

villebro commented on superset/connectors/druid/views.py:
```diff
@@ -49,7 +51,18 @@
 
 
-class DruidColumnInlineView(CompactCRUDMixin, SupersetModelView):
+class EnsureEnabled:
```
nit: could this be called `EnsureEnabledMixin`?

villebro commented on superset/connectors/druid/views.py:
```diff
@@ -49,7 +51,18 @@
+            raise NotFound()
+
+
+class DruidColumnInlineView(CompactCRUDMixin, SupersetModelView, EnsureEnabled):
```
nit: while it really doesn't matter here, `EnsureEnabled` should probably come to the left of `SupersetModelView` as per convention (same for the others).

villebro commented on superset/connectors/sqla/views.py:
```diff
@@ -369,6 +371,15 @@ class RowLevelSecurityFiltersModelView(  # pylint: disable=too-many-ancestors
         add_form_query_rel_fields = app.config["RLS_FORM_QUERY_REL_FIELDS"]
         edit_form_query_rel_fields = add_form_query_rel_fields
 
+    @staticmethod
+    def is_enabled() -> bool:
+        return is_feature_enabled("ROW_LEVEL_SECURITY")
+
+    @before_request
+    def ensure_enabled(self) -> None:
+        if not self.is_enabled():
+            raise NotFound()
```
We could potentially generalize this by breaking this out into a Mixin that checks `is_feature_enabled` on a reserved property. Something like this:

```python
class FeatureFlagMixin:
    feature_flag: str

    def is_enabled(self) -> bool:
        return is_feature_enabled(self.feature_flag)

    @before_request
    def ensure_enabled(self) -> None:
        if not self.is_enabled():
            raise NotFound()
```

Then later this could be used as follows:

```python
class RowLevelSecurityFiltersModelView(  # pylint: disable=too-many-ancestors
    FeatureFlagMixin, SupersetModelView, DeleteMixin
):
    feature_flag = "ROW_LEVEL_SECURITY"
    ....
```
we could potentially also add support for callbacks that make it possible to add custom validators that don't rely on `is_feature_enabled`:

```python
class FeatureFlagMixin:
    feature_flag: Union[str, Callable[[], str]]

    def is_enabled(self) -> bool:
        if callable(self.feature_flag):
            return self.feature_flag()
        return is_feature_enabled(self.feature_flag)

    @before_request
    def ensure_enabled(self) -> None:
        if not self.is_enabled():
            raise NotFound()
```

and then use it like this:

```python
    feature_flag = lambda _: app.config["FAB_ADD_SECURITY_VIEWS"] and app.config["SUPERSET_LOG_VIEW"]
```

benjreinhart replied:
I think we could, but it's a little bit more complicated then that. `@before_request` needs to be able to be scoped to a subset of route handlers in some cases. We can achieve that, but for now I think we should ship the more naive version. If this pattern continues to be used moving forward, the opportunity for a refactor should present itself.

villebro replied:
Gotcha - I don't have the full picture of the context in which this change is used, so some of my comments will probably be irrelevant 👍

villebro reviewed (commented):
Exciting stuff! Left a few comments/ideas

benjreinhart committed:
Better naming and follow conventions

benjreinhart commented:
@villebro Thank you for the solid feedback as always! I updated the naming and moved the order of inheritance around per the convention. If you're ok with it, I think my preference would be to hold off on a refactor around the mixin for now. Let me know what you think.

robdiciuccio commented on tests/alerts_tests.py:
```diff
@@ -359,3 +364,47 @@ def test_deliver_alert_screenshot(
+            self.login("admin")
+            uri = "/alertlogmodelview/list/"
+            rv = self.client.get(uri)
+            self.assertNotEqual(rv.status_code, 404)
```
Nit: might be better to do a positive assertion (200) on these.

benjreinhart replied:
Yes I agree, unfortunately many of these return 302s since I believe more setup is required to get the right users access to the resources. Ideally, these tests would do that, but seeing as I wasn't able to resolve it right away, I wanted to get this PR out and also keep the changes to a minimum.

robdiciuccio replied:
How about `self.assertLess(rv.status_code, 400)` in that case?

benjreinhart replied:
That works for me 👍🏻

benjreinhart committed:
Use assertLess over assertNotEqual

benjreinhart committed:
Assert less than 400

villebro reviewed (approved):
LGTM 👍

dpgaspar reviewed (commented):
LGTM, after tests pass

benjreinhart commented on superset/app.py:
```diff
@@ -226,10 +226,12 @@ def init_views(self) -> None:
+            "Home",
+            label=__("Home"),
+            href="/superset/welcome/",
+            cond=lambda: bool(appbuilder.app.config["LOGO_TARGET_PATH"]),
```
I'm not sure, but this should maintain parity with the previous code.

benjreinhart committed:
Fix failing test

vedangparasnis commented:
What will be the json payload format required for rowlevelsecurityfiltersmodelview/api/create . I can see there is an api for rowlevelsecurityfiltersmodelview/api/, however, can this api be used to create a new row level security using api. 

Thanks in advance
